[FLINK-20352][docs] Updated CLI documentation.

XComp opened the pull request:
## What is the purpose of the change

The CLI documentation was updated in the context of the efforts around FLINK-20341.
![cli](https://user-images.githubusercontent.com/1101012/100884603-f8070f00-34b1-11eb-9c6d-c43dd9cbe77a.png)

XComp commented:
@zentol @rmetzger @tillrohrmann first version of the CLI docs is ready. Looking forward to your feedback.

XComp commented:
@kl0u it would be nice to get some feedback from you as well

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,254 @@ specific language governing permissions and limitations
+or [YARN]({% link deployment/resource-providers/yarn.md %}) which are described in more detail in the 
+Resource Provider section. You can specify the mode you're Flink cluster is running in through the 
+`--target` parameter. It supports the following values:
+* `local`
+* `remote`
+* `kubernetes-session`
+* `kubernetes-application`
+* `yarn-per-job`
+* `yarn-session`
+* `yarn-application`
```
I think it would be good to document that `*-application` can only be used with `run-application` and the others only with `run`.

What is actually the difference between `flink list --target yarn-per-job` and `flink list --target yarn-session`? I guess one need to specify somehow that one wants to talk to a Yarn cluster. But the whether it is a session, a per-job or a application cluster should not really matter. @kl0u can you maybe chime in here?

kl0u replied:
There is no difference between the two commands that you mentioned @tillrohrmann . Both tell the CLI that it will be "talking" to a YARN cluster so it will go essentially through the `YarnClusterDescriptor`. 

As for documenting that each of the `run` and `run-application` are compatible with a subset with the available `execution.targets`, I agree.

XComp replied:
@kl0u Could you come up with a proposal for the `Selecting Deployment Targets` section? 😇 You might have more insights it might want to add a bit more? Or is the action-to-exection-target mapping the only important information that should be added to this section?

tillrohrmann replied:
Yes, it would be great to add some more information to this section.

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,254 @@ specific language governing permissions and limitations
 
-Disposes the savepoint at the given path. The savepoint path is returned by the savepoint trigger command.
-
-If you use custom state instances (for example custom reducing state or RocksDB state), you have to specify the path to the program JAR with which the savepoint was triggered in order to dispose the savepoint with the user code class loader:
```
Why did we drop this information? Is this no longer valid or too expert usage?

XComp replied:
Good catch 👍

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,254 @@ specific language governing permissions and limitations
 
-The run command has a savepoint flag to submit a job, which restores its state from a savepoint. The savepoint path is returned by the savepoint trigger command.
-
-By default, we try to match all savepoint state to the job being submitted. If you want to allow to skip savepoint state that cannot be restored with the new job you can set the `allowNonRestoredState` flag. You need to allow this if you removed an operator from your program that was part of the program when the savepoint was triggered and you still want to use the savepoint.
```
This bit `allowNonRestoredState` is also an interesting information for users.

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,254 @@ specific language governing permissions and limitations
-A "stop" call is a more graceful way of stopping a running streaming job, as the "stop" signal flows from
-source to sink. When the user requests to stop a job, all sources will be requested to send the last checkpoint barrier
-that will trigger a savepoint, and after the successful completion of that savepoint, they will finish by calling their
-`cancel()` method. If the `-d` flag is specified, then a `MAX_WATERMARK` will be emitted before the last checkpoint
-barrier. This will result all registered event-time timers to fire, thus flushing out any state that is waiting for
-a specific watermark, e.g. windows. The job will keep running until all sources properly shut down. This allows the
- job to finish processing all in-flight data.
```
The `-d` option sounds important as well.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @XComp. I think the updated documentation goes in very good direction. I like it. I had a couple of comments concerning some left out information. Moreover, I think that we probably should better document the different target values and in which combination they are valid. Maybe @kl0u can help us with this.

kl0u commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,254 @@ specific language governing permissions and limitations
-Use the `stop` to gracefully stop a running streaming job with a savepoint.
+### Creating a Savepoint
 
+Savepoints can be created to save the current state a job is in. All that's needed is the JobID:
```
What about adding a link to the "Savepoints" page?

kl0u commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,254 @@ specific language governing permissions and limitations
-By default, we try to match all savepoint state to the job being submitted. If you want to allow to skip savepoint state that cannot be restored with the new job you can set the `allowNonRestoredState` flag. You need to allow this if you removed an operator from your program that was part of the program when the savepoint was triggered and you still want to use the savepoint.
+#### Stop: Cancelling a Job but creating a final Savepoint
 
+Another action for stopping a job is `stop`. It function in the same way as `cancel` but creates a 
```
"It function" -> "It functions"

kl0u commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,254 @@ specific language governing permissions and limitations
-The run command has a savepoint flag to submit a job, which restores its state from a savepoint. The savepoint path is returned by the savepoint trigger command.
-
-By default, we try to match all savepoint state to the job being submitted. If you want to allow to skip savepoint state that cannot be restored with the new job you can set the `allowNonRestoredState` flag. You need to allow this if you removed an operator from your program that was part of the program when the savepoint was triggered and you still want to use the savepoint.
+#### Stop: Cancelling a Job but creating a final Savepoint
```
Here there is also the DRAIN option where we also emit the `MAX_WATERMARK`, right?

kl0u commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,254 @@ specific language governing permissions and limitations
+
+Flink supports different deployment backends like [Kubernetes]({% link deployment/resource-providers/native_kubernetes.md %}) 
+or [YARN]({% link deployment/resource-providers/yarn.md %}) which are described in more detail in the 
+Resource Provider section. You can specify the mode you're Flink cluster is running in through the 
```
Link to the Resource Provider Section?

XComp replied:
We don't have a dedicated resource provider section. That's why I mentioned two examples to be able to at least reference to their subsections.

kl0u reviewed (commented):
I also like the changes. The page now looks more readable. I left some minor comments.

tillrohrmann commented:
Another thing I stumbled across is the deployment example for a Python job. Maybe we add it here or we create a dedicated page for how to interact with Python programs. The latter is probably the better solution.

XComp commented:
Thanks for looking into the changes @kl0u & @tillrohrmann . I addressed your comments. Feel free to check out the new version.
![cli-v2](https://user-images.githubusercontent.com/1101012/101410465-a90d0f80-38df-11eb-8a87-fbeb41752289.png)

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,303 @@ specific language governing permissions and limitations
+Another action for stopping a job is `stop`. It is a more graceful way of stopping a running streaming 
+job as the `stop`  flows from source to sink. When the user requests to stop a job, all sources will 
+be requested to send the last checkpoint barrier that will trigger a savepoint, and after the successful 
+completion of that savepoint, they will finish by calling their	`cancel()` method. 
```
```suggestion
completion of that savepoint, they will finish by calling their `cancel()` method. 
```

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,303 @@ specific language governing permissions and limitations
 
-The run command has a savepoint flag to submit a job, which restores its state from a savepoint. The savepoint path is returned by the savepoint trigger command.
+If the `--drain` flag is specified, then a `MAX_WATERMARK` will be emitted before the last checkpoint 
+barrier. This will result all registered event-time timers to fire, thus flushing out any state that 
```
```suggestion
barrier. This will make all registered event-time timers to fire, thus flushing out any state that 
```

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,303 @@ specific language governing permissions and limitations
-Disposes the savepoint at the given path. The savepoint path is returned by the savepoint trigger command.
+See how the command is equal to the [initial run command](#submitting-a-job) except for the 
+`--fromSavepoint` parameter which is used to refer to the state of the 
+[previously cancelled job](#stopping-a-job-gracefully-creating-a-final-savepoint). A new JobID is 
```
```suggestion
[previously stopped job](#stopping-a-job-gracefully-creating-a-final-savepoint). A new JobID is 
```

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,303 @@ specific language governing permissions and limitations
+            </td>
+        </tr>
+        <tr>
+            <td><code class="highlighter-rouge">info</code></td>
```
This command hasn't been mentioned explicitly in the sections above. Was this intentional?

XComp replied:
I considered the `info` action being for more advanced users. The CLI job lifecycle management addresses beginners instead. That was the reason I didn't add it to the `Job Lifecycle Management` section.

tillrohrmann replied:
Ok, sounds good to me. I think that this command is also not used very often.

tillrohrmann commented on docs/deployment/cli.md:
```diff
@@ -23,664 +23,303 @@ specific language governing permissions and limitations
+
+Flink supports different deployment backends like [Kubernetes]({% link deployment/resource-providers/native_kubernetes.md %}) 
+or [YARN]({% link deployment/resource-providers/yarn.md %}) which are described in more detail in the 
+Resource Provider section. You can specify the mode you're Flink cluster is running in through the 
```
```suggestion
Resource Provider section. You can specify the mode your Flink cluster is running in through the 
```

tillrohrmann reviewed (approved):
Thanks for updating this PR @XComp. I had a few minor comments. After more information for the deployment targets are added, I am +1 for merging this PR.

tillrohrmann reviewed (approved):
Thanks for updating this PR @XComp and your help @kl0u. The changes look good to me. Merging this PR now.
